File: easy/move-zeroes/python/move_zeroes.py
from typing import List


def move_zeroes(nums: List[int]) -> List[int]:
    """
    Moves all zeros to the end while maintaining the order of non-zero elements.
    """
    pos = 0
    for i in range(len(nums)):
        if nums[i] != 0:
            nums[pos], nums[i] = nums[i], nums[pos]
            pos += 1
    return nums


if __name__ == "__main__":
    tests = [
        [0, 1, 0, 3, 12],
        [0, 0, 1],
        [4, 5, 6],
        [0, 0, 0],
        [1, 0, 2, 0, 3, 0],
    ]

    for t in tests:
        print(f"Input: {t} → Output: {move_zeroes(t[:])}")

File: easy/move-zeroes/cpp/move_zeroes.cpp
#include <iostream>
#include <vector>
using namespace std;

void moveZeroes(vector<int>& nums) {
    int nonZeroPos = 0;
    
    // Move all non-zero elements to the front
    for (int i = 0; i < (int)nums.size(); ++i) {
        if (nums[i] != 0) {
            nums[nonZeroPos] = nums[i];
            ++nonZeroPos;
        }
    }
    
    // Fill remaining positions with zeros
    for (int i = nonZeroPos; i < (int)nums.size(); ++i) {
        nums[i] = 0;
    }
}

void printVector(const vector<int>& nums) {
    cout << "[";
    for (int i = 0; i < (int)nums.size(); ++i) {
        cout << nums[i];
        if (i < (int)nums.size() - 1) cout << ",";
    }
    cout << "]" << endl;
}

int main() {
    // Test cases
    vector<int> test1 = {0, 1, 0, 3, 12};
    moveZeroes(test1);
    printVector(test1); // [1,3,12,0,0]
    
    vector<int> test2 = {0, 0, 1};
    moveZeroes(test2);
    printVector(test2); // [1,0,0]
    
    vector<int> test3 = {4, 5, 6};
    moveZeroes(test3);
    printVector(test3); // [4,5,6]
    
    vector<int> test4 = {0, 0, 0};
    moveZeroes(test4);
    printVector(test4); // [0,0,0]
    
    vector<int> test5 = {1, 0, 2, 0, 3, 0};
    moveZeroes(test5);
    printVector(test5); // [1,2,3,0,0,0]
    
    return 0;
}
